src/Features/CSharp/Portable/CodeRefactorings/InlineMethod/CSharpInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,167 @@
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.InlineMethod;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.InlineMethod
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(PredefinedCodeRefactoringProviderNames.InlineMethod)), Shared]
+    [Export(typeof(CSharpInlineMethodRefactoringProvider))]
+    internal sealed class CSharpInlineMethodRefactoringProvider :
+        AbstractInlineMethodRefactoringProvider<BaseMethodDeclarationSyntax, StatementSyntax, ExpressionSyntax, InvocationExpressionSyntax>
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpInlineMethodRefactoringProvider()
+            : base(CSharpSyntaxFacts.Instance, CSharpSemanticFactsService.Instance)
+        {
+        }
+
+        protected override ExpressionSyntax? GetRawInlineExpression(BaseMethodDeclarationSyntax methodDeclarationSyntax)
+        {
+            var blockSyntaxNode = methodDeclarationSyntax.Body;
+            if (blockSyntaxNode != null)
+            {
+                // 1. If it is an ordinary method with block
+                var blockStatements = blockSyntaxNode.Statements;
+                if (blockStatements.Count == 1)
+                {
+                    var statementSyntax = blockStatements[0];
+                    return statementSyntax switch
+                    {
+                        // Note: For this case this will return null in Callee()
+                        // void Caller() { Callee(); }
+                        // void Callee() { return; }
+                        // Refactoring won't be provided for this case.
+                        ReturnStatementSyntax returnStatementSyntax => returnStatementSyntax.Expression,
+                        ExpressionStatementSyntax expressionStatementSyntax => expressionStatementSyntax.Expression,
+                        ThrowStatementSyntax throwStatementSyntax => throwStatementSyntax.Expression,
+                        _ => null
+                    };
+                }
+            }
+            else
+            {
+                // 2. If it is an Arrow Expression
+                var arrowExpressionNode = methodDeclarationSyntax.ExpressionBody;
+                return arrowExpressionNode?.Expression;
+            }
+
+            return null;
+        }
+
+        protected override SyntaxNode? GetEnclosingMethodLikeNode(SyntaxNode syntaxNode)
+        {
+            for (var calleeNode = syntaxNode; calleeNode != null; calleeNode = calleeNode.Parent)
+            {
+                // 1. Common method
+                // 2. Local function
+                // 3. Simple lambda & Parenthesized Lambda
+                // 4. Anonymous Method ExpressionSyntax
+                if (calleeNode is BaseMethodDeclarationSyntax
```

Comment:
is this ok if you're parented by a constructor/accessor/operator/etc.?  or is this codepath not called in that case?

can you doc if this is/isn't possible.  if it is possible, doc why null is ok in those cases.

Reply:
I think about this function for a while, it is only used for this case https://github.com/Cosifne/roslyn/blob/d5dd0ec699f6b07e545afa2989e9dabadd2e1f2e/src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.InlineContext.cs#L283
to check 
`Func<Task> Callee() => async () => await Task.CompltedTask `. the 'await' is not an 'await' in the callee.
But since we could filter the lambda case before goes to all the childern. This might not be useful acutally.
It is true it is only called for 
1. Constructor
2. Common method
3. Desconstructor (I guess this could be private)